terraform/mssql/provision/data.tf

```diff
@@ -1,14 +1,83 @@
 locals {
-  port = 1433
+  port               = 1433
+  vpc_id             = length(var.aws_vpc_id) > 0 ? data.aws_vpc.provided[0].id : data.aws_vpc.default.id
+  subnet_ids         = length(var.rds_subnet_group) > 0 ? data.aws_subnets.in_subnet_group.ids : data.aws_subnets.in_vpc.ids
+  security_group_ids = length(var.rds_vpc_security_group_ids) > 0 ? data.aws_security_groups.provided[0].ids : [aws_security_group.rds-sg[0].id]
+  subnet_group_name  = length(var.rds_subnet_group) > 0 ? data.aws_db_subnet_group.provided[0].name : aws_db_subnet_group.rds-private-subnet[0].name
 }
 
-data "aws_vpc" "vpc" {
+data "aws_vpc" "default" {
   default = true
 }
 
-data "aws_subnets" "all" {
+data "aws_subnets" "in_vpc" {
   filter {
     name   = "vpc-id"
-    values = [data.aws_vpc.vpc.id]
+    values = [local.vpc_id]
+  }
+}
+
+data "aws_vpc" "provided" {
+  count = length(var.aws_vpc_id) > 0 ? 1 : 0
+  id    = var.aws_vpc_id
+
+  lifecycle {
+    postcondition {
+      condition     = length(var.aws_vpc_id) == 0 || var.aws_vpc_id == self.id
+      error_message = "the specified vpc doesn't exist"
+    }
+  }
+}
+
+data "aws_db_subnet_group" "provided" {
+  count = length(var.rds_subnet_group) > 0 ? 1 : 0
+  name  = var.rds_subnet_group
+
+  lifecycle {
+    postcondition {
+      condition     = length(var.aws_vpc_id) == 0 || var.aws_vpc_id == self.vpc_id
+      error_message = "the specified subnet group doesn't exist or doesn't correspond to the specified vpc"
```

Comment:
The error message should state that it doesn't correspond to the VPC. If it didn't exist, terraform will fail before reaching this postcondition with the error that was tested in the terraform-tests. The scenario where the postcondition is executed remains untested. I think it does provide value, but we should think of ways of testing it.

Reply:
Data sources logic has been improved in latest changes.